ledger/participant-integration-api/src/main/scala/platform/apiserver/services/tracking/TrackerMap.scala

```diff
@@ -59,32 +62,44 @@ private[services] final class TrackerMap[Key](
       .getOrElse(
         key,
         lock.synchronized {
-          trackerBySubmitter.getOrElse(
-            key, {
-              val r = new TrackerMap.AsyncResource(newTracker(key).map { t =>
-                logger.info(s"Registered tracker for submitter $key")
-                new TrackerWithLastSubmission(t)
-              })
-              trackerBySubmitter += key -> r
-              r
-            },
-          )
+          trackerBySubmitter.getOrElse(key, registerNewTracker(key))
         },
       )
-      .flatMap(_.track(request))
+      .withResource(_.track(request))
+  }
+
+  private def registerNewTracker(
+      key: Key
+  )(implicit executionContext: ExecutionContext): AsyncResource[TrackerWithLastSubmission] = {
+    val resource = new AsyncResource(
+      newTracker(key)
+        .andThen {
+          case Success(_) =>
+            logger.info(s"Registered a tracker for submitter $key.")
+          case Failure(exception) =>
+            logger.error("Failed to register a tracker.", exception)
+        }
+        .map(new TrackerWithLastSubmission(_))
+    )
+    trackerBySubmitter += key -> resource
+    resource
   }
 
   def cleanup(): Unit = lock.synchronized {
-    val nanoTime = System.nanoTime()
-    trackerBySubmitter foreach { case (submitter, trackerResource) =>
-      trackerResource.ifPresent { tracker =>
-        if (nanoTime - tracker.getLastSubmission > retentionNanos) {
-          logger.info(
-            s"Shutting down tracker for $submitter after inactivity of $retentionPeriod"
-          )(trackerResource.loggingContext)
+    val currentTime = System.nanoTime()
+    trackerBySubmitter.foreach { case (submitter, trackerResource) =>
+      trackerResource.currentState match {
+        case Ready(tracker) =>
+          if (currentTime - tracker.getLastSubmission > retentionNanos) {
+            logger.info(
+              s"Shutting down tracker for $submitter after inactivity of $retentionPeriod"
+            )(trackerResource.loggingContext)
+            trackerBySubmitter -= submitter
+            tracker.close()
+          }
+        case Waiting(_) => // do nothing
```

Comment:
Shouldn't we call close here?

Reply:
This is the cleanup function. It should only close stuff that's expired. If it's still waiting, it's not expired.

Reply:
Alright got it. So the retention is only for ready trackers. Can we update the comment 
`  * @param retentionPeriod The minimum finite duration for which to retain idle trackers.` to have the same terminology

Reply:
Can you be clear about what you'd like to change? I don't understand how this comment and the behavior are misaligned.

Reply:
We're implicitly saying `ready` = `idle` so I would propose updating the comment to `...for which to retain ready trackers`

Reply:
Updated a bunch of comments to make this clearer.